core/translate/main_loop.rs

```
@@ -743,23 +773,205 @@ pub fn open_loop(
+                    if let Table::BTree(btree) = &build_table.table {
+                        program.emit_insn(Insn::OpenRead {
+                            cursor_id,
+                            root_page: btree.root_page,
+                            db: build_table.database_id,
+                        });
+                    }
+
+                    cursor_id
+                };
+
+                // Allocate registers for hash table and join keys
+                let hash_table_reg = program.alloc_register();
+                let num_keys = hash_join_op.join_key_columns.len();
+                let build_key_start_reg = program.alloc_registers(num_keys);
+
+                // Store column indices for join keys
+                for (idx, (build_col, _probe_col)) in
+                    hash_join_op.join_key_columns.iter().enumerate()
+                {
+                    program.emit_insn(Insn::Integer {
+                        value: *build_col as i64,
+                        dest: build_key_start_reg + idx,
+                    });
+                }
+
+                // Create new loop for hash table build phase
+                let build_loop_start = program.allocate_label();
+                let build_loop_end = program.allocate_label();
+                let skip_to_next = program.allocate_label();
+
+                program.emit_insn(Insn::Rewind {
+                    cursor_id: build_cursor_id,
+                    pc_if_empty: build_loop_end,
+                });
+
+                program.preassign_label_to_next_insn(build_loop_start);
+
+                // Apply WHERE clause filters that reference only the build table
+                // Rows that don't match are skipped (not added to hash table)
+                let build_table_join_index = join_order
+                    .iter()
+                    .position(|j| j.original_idx == hash_join_op.build_table_idx);
+                if let Some(build_join_idx) = build_table_join_index {
```

silently ignoring case where this is None?